### PR TITLE
feat: Adding skipSchemaValidation flag (#20771)

### LINKED ISSUE #20771: Add support for passing --skip-schema-validation to help
# Summary

Add the ability to skip helm schema validation in helm template (pass `--skip-schema-validation` when argo uses `helm template`)

# Motivation

When running offline clusters, using helm charts that refer to external schemas for validation is very challenging. Projects historically have had to chose between chose between providing a schema for those with internet connectivity (usually to something like kubernetesjsonschema.dev) to aid in debugging and breaking those without an internet connection.

Since Helm 3.16, helm has a new flag for `helm template`, `--skip-schema-validation` that can be used to fix this issue for those of us that have clusters with no internet connectivity.

# Proposal

The most straightforward way to implement this would be to create a new property in the `Application` CRD that would be modelled after the `skipCRDs` property that passes `--skip-schema-validation` to the helm template command.

Alternatively, there could be a list property that allows passing arbitrary helm template parameters to the command. This would be more flexible, but might come with some around passing either flags that are controlled in other ways (like the skipCRDs flag).

### PR DESCRIPTION
**Pull request by dmosesson**
Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [X] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [X] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [X] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

Closes https://github.com/argoproj/argo-cd/issues/20771

Tentatively also closes:
* https://github.com/argoproj/argo-cd/issues/20771
* https://github.com/argoproj/argo-cd/pull/20678
* https://github.com/argoproj/argo-cd/pull/20650
* https://github.com/argoproj/argo-cd/issues/20287

This allows argo to deploy helm charts that have JSON schema files for helm charts in locations that do not have internet access.

Some notes:
* I had some issues regenerating some of the files, and probably too many of those are included (I tried to run `make codegen` and had issues, and `make codegen-local` ran with some slightly different versions I suspect.
* The kustomize tests for me mostly failed, but I think that is more an artifact of the above, where I installed in somewhat incorrectly
* Helm had to be updated to 3.16.0 for this change, not sure if that should be a separate PR or not (and most probably I missed some of the places it should have been updated)

### CODE REVIEW

File: docs/user-guide/helm.md [hunk 1]
```diff
@@ -499,6 +499,26 @@ spec:
       skipCrds: true
 ```
 
+## Helm `--skip-schema-validation`
+
+Helm validations the values.yaml file using a values.schema.json file. See [Schema files](https://helm.sh/docs/topics/charts/#schema-files) for details.
```
Comment (crenshaw-dev) [inline docs/user-guide/helm.md:504]:
```suggestion
Helm validates the values.yaml file using a values.schema.json file. See [Schema files](https://helm.sh/docs/topics/charts/#schema-files) for details.
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Besides one tweak, lgtm! Can you merge master and re-run codegen?

File: docs/user-guide/helm.md (hunk 1 above)
Comment (dmosesson) [reply docs/user-guide/helm.md:504]: I made the change, but when I try to run `make codegen` I get an error basically as if `make` or something that `make` is running is confined in some way? The errors are things like
* mkdir: cannot create directory '/go/src/github.com/gogo': Permission denied
* mkdir: cannot create directory '/go/src/k8s.io': Permission denied

And the project root is seemingly wrong as well? 
```
+++ dirname ./hack/generate-proto.sh
++ cd ./hack/..
++ pwd
+ PROJECT_ROOT=/go/src/github.com/argoproj/argo-cd
```

Running `make protogen-fast` by itself (the stage that is throwing the error) seems to work just fine though. Apologies if I am missing something basic, this is my first PR to argo.

Comment (dmosesson) [reply docs/user-guide/helm.md]: I was able to get the `make codegen` to run by forcing the `run-in-test-client` part to run as root in the container. This is not part of the PR, and the file generated so *shrug*? I think it should work at this point. The comment about fixing the help typo has been addressed

File: reposerver/repository/repository_test.go [hunk 2]
```diff
@@ -4325,3 +4327,45 @@ images:
 		}, res.Commands)
 	})
 }
+
+func Test_SkipSchemaValidation(t *testing.T) {
+	t.Run("helm", func(t *testing.T) {
+		service := newService(t, "testdata/broken-schema-verification")
+
+		// Fill the manifest request with as many parameters affecting Helm commands as possible.
```
Comment (crenshaw-dev) [inline reposerver/repository/repository_test.go:4335]:
```suggestion
```

I think this comment isn't relevant here.

File: reposerver/repository/repository_test.go [hunk 3]
```diff
@@ -4325,3 +4327,45 @@ images:
...
+
+		res, err := service.GenerateManifest(context.Background(), &q)
+
+		require.NoError(t, err)
+		assert.Equal(t, []string{"helm template . --name-template test-app --include-crds --skip-schema-validation"}, res.Commands)
+	})
+	t.Run("helm", func(t *testing.T) {
+		service := newService(t, "testdata/broken-schema-verification")
+
+		// Fill the manifest request with as many parameters affecting Helm commands as possible.
```
Comment (crenshaw-dev) [inline reposerver/repository/repository_test.go:4355]:
```suggestion
```

I think this comment isn't relevant here.

**Review by crenshaw-dev — DISMISSED** [review]
Nice test! One tweak

File: reposerver/repository/repository_test.go (hunk 3 above)
Comment (dmosesson) [reply reposerver/repository/repository_test.go]: Removed

**Comment by dmosesson** [comment]
@crenshaw-dev Are there any more changes you are looking for, or its good to go from your perspective? I think all the issues flagged above have been addressed

File: test/e2e/testdata/helm-repo/local/index.yaml [hunk 4]
```diff
@@ -5,3 +5,4 @@ entries:
       urls:
         - http://127.0.0.1:9080/argo-e2e/testdata.git/helm-repo/helm-1.0.0.tgz
       version: 1.0.0
+      name: helm
```
Comment (ishitasequeira) [inline test/e2e/testdata/helm-repo/local/index.yaml:8]: QQ: Is the name field required here? I do not think this change is related to `skipSchemaValidation` flag changes.

**Review by ishitasequeira — COMMENTED** [review]
Overall PR looks good to me!! Just left a quick question for understanding the extra change.

@crenshaw-dev any more concerns about the PR?

File: test/e2e/testdata/helm-repo/local/index.yaml (hunk 4 above)
Comment (dmosesson) [reply test/e2e/testdata/helm-repo/local/index.yaml:8]: It is. In helm 3.16, this property is now required (its not documented anywhere, but the test errors were saying it was a missing required attribute). Given that helm 3.16 is a requirement for this feature, it was added here

**Review by rumstead — APPROVED** [review]
LGTM

**Comment by crenshaw-dev** [comment]
@ishitasequeira no concerns from me

**Review by ishitasequeira — APPROVED** [review]
Changes look good!! Thanks @dmosesson !!

**Comment by crenshaw-dev** [comment]
@jdvorak8660 this should work fine for multi-source apps. The `helm.skipSchemaValidation` field is embedded in each item in the `spec.sources` field. They're the same struct.

**Comment by jdvorak8660** [comment]
> @jdvorak8660 this should work fine for multi-source apps. The `helm.skipSchemaValidation` field is embedded in each item in the `spec.sources` field. They're the same struct.

Thanks for the response.  I did not experience this when I attempted to use the option for multiple sources, but I'll go back and make sure I wasn't making some silly mistake.  When it didn't work as expected I assumed it was not an option as it is also not referenced in the documentation for multiple sources.

Any chance you have an example of a working multi source application with this option I could reference?

**Pull request merged**